src/mlio/image_reader.cxx

```diff
@@ -0,0 +1,224 @@
+{
+    std::vector<feature_desc> descs{};
+    // schema follows the NHWC convention
+    descs.emplace_back(
+        feature_desc_builder{"value",
+                             data_type::uint8,
+                             {params().batch_size,
+                              static_cast<std::size_t>(params_.data_shape[1]),
+                              static_cast<std::size_t>(params_.data_shape[2]),
+                              static_cast<std::size_t>(params_.data_shape[0])}}
+            .build());
+    schema_ = make_intrusive<schema>(std::move(descs));
+}
+
+intrusive_ptr<example>
+image_reader::decode(instance_batch const &batch) const
+{
+    auto tsr = make_tensor(batch.instances(), batch.size());
+    std::vector<intrusive_ptr<tensor>> tensors{};
+    tensors.emplace_back(tsr);
+    return make_intrusive<example>(schema_, std::move(tensors));
+}
+
+intrusive_ptr<dense_tensor>
+image_reader::make_tensor(std::vector<instance> const &instances,
+                          std::size_t batch_size) const
+{
+    auto instance_offset =
+        params_.data_shape[0] * params_.data_shape[1] * params_.data_shape[2];
+    auto total_memory_buffer =
+        static_cast<unsigned long>(instance_offset) * batch_size;
+    std::vector<std::uint8_t> img_vec(total_memory_buffer);
+    auto *ptr_img_vec = img_vec.data();
+    int type;
+    cv::ImreadModes mode;
+    for (const instance &ins : instances) {
+        memory_slice img_buf;
+        if (params_.img_frame == image_frame::pipe) {
+            img_buf = ins.bits().subslice(recordio_image_header_offset_);
+        }
+        else {
+            img_buf = ins.bits();
+        }
+        cv::Mat mat(
+            1,
+            static_cast<int>(img_buf.size()),
+            CV_8U,
+            static_cast<void *>(const_cast<std::byte *>(img_buf.data())));
+        switch (params_.data_shape[0]) {
+        case 1:
+            mode = cv::ImreadModes::IMREAD_GRAYSCALE;
+            type = CV_8UC1;
+            break;
+        case 3:
+            mode = cv::ImreadModes::IMREAD_COLOR;
+            type = CV_8UC3;
+            break;
+        case 4:
+            mode = cv::ImreadModes::IMREAD_UNCHANGED;
+            type = CV_8UC4;
+            break;
+        default:
+            throw std::invalid_argument{
+                fmt::format("Unsupported number of channels: {0:n}",
+                            params_.data_shape[0])};
+        }
+        cv::Mat tmp = decode_image(mat, mode, ins);
+        if (params_.resize != -1) {
+            resize(tmp, tmp);
+        }
+        if (params_.to_rgb && tmp.channels() != 1) {
+            cv::cvtColor(tmp, tmp, cv::COLOR_BGR2RGB);
+        }
+        cv::Mat dst{
+            params_.data_shape[1], params_.data_shape[2], type, ptr_img_vec};
+        crop(tmp, dst);
+        ptr_img_vec += instance_offset;
+    }
+
+    auto ptr = wrap_cpu_array<data_type::uint8>(std::move(img_vec));
+    auto tsr = make_intrusive<dense_tensor>(
```

Comment:
Is it pretty standard across ML-IO to call the returned tensor `tsr`?

Reply:
`tsr` seems to be used to refer to a tensor at various places. However, we can refactor this if this sounds confusing to you

Reply:
No need, just wanted to clarify this was the case.